pkg/election/election.go

```diff
@@ -255,67 +263,60 @@ func (e *Election) campaignLoop(ctx context.Context, session *concurrency.Sessio
 
 				e.l.Info("get response from election observe", zap.String("key", string(resp.Kvs[0].Key)), zap.String("value", string(resp.Kvs[0].Value)))
 				leaderKey = string(resp.Kvs[0].Key)
-				leaderInfo, err := getCompaingerInfo(resp.Kvs[0].Value)
+				leaderInfo, err = getCompaingerInfo(resp.Kvs[0].Value)
 				if err != nil {
 					// this should never happened
 					e.l.Error("fail to get leader information", zap.String("value", string(resp.Kvs[0].Value)), zap.Error(err))
 					continue
 				}
-				leaderID = leaderInfo.ID
 
-				if oldLeaderID == leaderID {
+				if oldLeaderID == leaderInfo.ID {
 					continue
 				} else {
-					oldLeaderID = leaderID
+					oldLeaderID = leaderInfo.ID
 					break observeElection
 				}
 			}
 		}
 
-		if len(leaderID) == 0 {
+		if leaderInfo == nil || len(leaderInfo.ID) == 0 {
 			cancel2()
 			compaignWg.Wait()
 			continue
 		}
 
-		if leaderID != e.info.ID {
-			e.l.Info("current member is not the leader", zap.String("current member", e.info.ID), zap.String("leader", leaderID))
-			e.isNotLeader()
+		if leaderInfo.ID != e.info.ID {
+			e.l.Info("current member is not the leader", zap.String("current member", e.info.ID), zap.String("leader", leaderInfo.ID))
```

Comment:
How about logging `e.info` and `leaderInfor` directly (including `ID` and `Addr`)? If so, ditto for other places.

Reply:
update in 2c624cc